c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+        mut rws: Vec<(S, Rewrite<S>)>,
+    ) -> (Vec<RewriteTree<S>>, Vec<(S, Rewrite<S>, RewriteError<S>)>) {
+        // Sort by start position and then by decreasing length, so that each parent span comes
+        // before all its children.
+        rws.sort_by_key(|&(ref s, _)| (s.lo(), Reverse(s.hi() - s.lo())));
+
+        // The stack contains a `RewriteTree` node for the most recently processed item from `rws`
+        // along with all of its ancestors (nodes whose spans strictly contain its span).
+        // `stack[0]` is the outermost ancestor, and `stack.last()` is the most recently pushed
+        // node.  Nodes are "committed" once they are known to have no more children; when
+        // committed, the node is moved into the `children` list of its parent or (if it has no
+        // parent) into the output list `out`.
+        let mut stack = Vec::<RewriteTree<S>>::new();
+        let mut out = Vec::new();
+        let mut errs = Vec::new();
+
+        fn commit<S: SpanLike>(
+            stack: &mut Vec<RewriteTree<S>>,
+            out: &mut Vec<RewriteTree<S>>,
+            rt: RewriteTree<S>,
+        ) {
+            if let Some(parent) = stack.last_mut() {
+                debug_assert!(parent.span.contains(rt.span));
+                debug_assert!(parent
+                    .children
+                    .last()
+                    .map_or(true, |ch| ch.span.hi() <= rt.span.lo()),);
+                parent.children.push(rt);
+            } else {
+                debug_assert!(out.last().map_or(true, |ch| ch.span.hi() <= rt.span.lo()),);
+                out.push(rt);
+            }
+        }
+
+        for (span, rw) in rws {
+            let lo = span.lo();
+
+            // Handle items with identical spans.
+            //
+            // If `rws` contains two items with the same span, they should be adjacent in the
+            // sorted list.  If the first item of such a run was pushed onto the stack, we will
+            // catch it here when processing the second item.  Since this case avoids pushing or
+            // committing any items, all remaining items in the run will land here too.
+            if let Some(other) = stack.last().filter(|other| other.span == span) {
+                if other.rw != rw {
+                    // This item has the same span as the previous one, but wants to perform a
+                    // different rewrite.
+                    errs.push((span, rw, RewriteError::Conflict));
+                }
+                continue;
+            }
+
+            // Commit all rewrites that come strictly before `span`.  This leaves only nodes that
+            // at least partially overlap the current item.  These should normally be ancestors of
+            // the current item on the stack, but it's also possible that the current item
+            // erroneously partially overlaps a previous item.
+            while stack.last().map_or(false, |rt| rt.span.hi() <= lo) {
```

Comment:
i think `ch.hi() <= parent.lo()` can be factored out into another method a la `contains` or `overlaps`, and would help make things clearer

Reply:
I think that's more trouble than it's worth.  There are only four places with similar comparisons (2x `a.hi() <= b.lo()`, 1x `a.hi() <= lo`, 1x `hi <= b.lo()` - hence the trouble), all four already have comments explaining the high-level logic, and it's not too hard to figure out the exact meaning of the comparison regardless (low vs high comparisons like this are fairly common when working with ranges, I think).

Note the existing `SpanLike` methods exactly match methods on the actual `Span` type, and `Span` doesn't have a method for this sort of comparison.

Reply:
Alright, I don't feel too strongly about this but maintain it would make things clearer.